File: #134/java/Solution.java
/*
 * https://leetcode.com/problems/gas-station/
 *Runtime: 0 ms, faster than 100.00% of Java online submissions for Gas Station.
 *Memory Usage: 37.5 MB, less than 98.04% of Java online submissions for Gas Station.
*/

class Solution
{
    int gas[];
    int cost[];
    
    public int canCompleteCircuit(int gas[], int cost[]) {
	int sumRemaining = 0; // track current remaining
	int total = 0; // track total remaining
	int start = 0; //starting
 
	for (int i = 0; i < gas.length; i++) {
		int remaining = gas[i] - cost[i];
 
		//if sum remaining of (i-1) >= 0, continue 
		if (sumRemaining >= 0) { 
			sumRemaining += remaining;
		//otherwise, reset start index to be current
		} else {
			sumRemaining = remaining;
			start = i;
		}
		total += remaining;
	}
 
	if (total >= 0){
		return start;
	}else{
		return -1;
	}
}
}

File: #134/cpp/Solution.cpp
/**
 * https://leetcode.com/problems/gas-station/
 * runtime: 4 ms, faster than 99.80% of C++ online submissions
 * Memory Usage: 10 MB, less than 27.25% of C++ online submissions
 */
 
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        for (int i = 0; i < gas.size(); ++i) {
            gas[i] -= cost[i];
        }
        int ans = 0;
        int sum = 0;
        for (int i = 0; i < gas.size()*2-1; ++i) {
            sum += gas[i % gas.size()];
            if (sum < 0) {
                ans = i + 1;
                if (ans >= gas.size()) {
                    return -1;
                }
                sum = 0;
            }
        }
        return ans;
    }
};
